Core: Handle various edge cases related to executable permissions.

StevenMiller123 opened the pull request:
As I tested various forms of protections on memory, I found a handful of inaccuracies this PR will address.
- Execute permissions are hidden from direct memory mappings.
- Attempting to map backed memory with execute permssions could fail on Windows, due to an edge case related to `MapViewOfFile3`. Fixing this is necessary for `sceKernelMprotect` to work properly on flexible memory after #3639 
- Protecting memory with read | execute (and no write permissions) would behave strangely on Windows. This was due to an issue with the flags we applied in `AddressSpace::Impl::Protect`, so I cleaned up the logic there too.
- Internally, the PS4 kernel seems to treat read and write permissions separately, but applies read permissions whenever only write is specified. Handling this properly fixes the raw value of VMA protections occasionally being exactly one less than equivalent mappings on real hardware.
- `sceKernelMapDirectMemory2` returns EINVAL when requesting execute permissions, not EACCES.

Commit:
Fix flag handling on Windows

Fixes a weird homebrew kalaposfos made

Commit:
Fix backing protects

Windows requires that protections on areas committed through MapViewOfFile functions are less than the original mapping.
The best way to make sure everything works is to VirtualProtect the code area with the requested protection instead of applying prot directly.

Commit:
Fix error code for sceKernelMapDirectMemory2

Real hardware returns EINVAL instead of EACCES here

Commit:
Fix prot setting in ProtectBytes

Commit:
Handle some extra protection-related edge cases.

Real hardware treats read and write as separate perms, but appends read if you call with write-only (this is visible in VirtualQuery calls)

Additionally, execute permissions are ignored when protecting dmem mappings.

Commit:
Properly handle exec permission behavior for memory pools

Calling sceKernelMemoryPoolCommit with executable permissions returns EINVAL, mprotect on pooled mappings ignores the exec protection.

Commit:
Clang

Commit:
Allow execution protection for direct memory

Further hardware tests show that the dmem area is actually executable, this permission is just hidden from the end user.

StevenMiller123 committed:
Clang

squidbus commented on src/core/address_space.cpp:
```diff
@@ -155,7 +155,12 @@ struct AddressSpace::Impl {
-                                     phys_addr, size, MEM_REPLACE_PLACEHOLDER, prot, nullptr, 0);
+                                     phys_addr, size, MEM_REPLACE_PLACEHOLDER,
+                                     PAGE_EXECUTE_READWRITE, nullptr, 0);
+                ASSERT_MSG(ptr, "{}", Common::GetLastErrorMsg());
```
Probably should have some kind of message with this similar to the one above, like `MapViewOfFile3 failed. {}` maybe.

StevenMiller123 committed:
More descriptive assert message

StevenMiller123 committed:
Align address and size in mmap

Like most POSIX functions, mmap aligns address down to the nearest page boundary, and aligns address up to the nearest page boundary.
Since mmap is the only memory mapping function that doesn't error early on misaligned length or size, handle the alignment in the libkernel code.

StevenMiller123 committed:
Clang

squidbus reviewed (approved):
Presumably needs testing but code looks good to me.

Commit:
Fix valid flags

After changing the value, games that specify just CpuWrite would hit the error return.

Commit:
Fix prot conversion functions

The True(bool) function returns true whenever value is greater than 0. While this rarely manifested before because of our wrongly defined CpuReadWrite prot, it's now causing trouble with the corrected values.
Technically this could've also caused trouble with games mapping GpuRead permissions, but that seems to be a rare enough use case that I guess it never happened?

I've also added a warning for the case where `write & !read`, since we don't properly handle write-only permissions, and I'm not entirely sure what it would take to deal with that.

Commit:
Fix some lingering dmem issues

ReleaseDirectMemory was always unmapping with the size parameter, which could cause it to unmap too much. Since multiple mappings can reference the same dmem area, I've calculated how much of each VMA we're supposed to unmap.
Additionally, I've adjusted the logic for carving out the free dmem area to properly work if ReleaseDirectMemory is called over multiple dmem areas.

Finally, I've patched a bug with my code in UnmapMemory.
